File: .pnpm-store/v11/projects/e6a985c005a1213f9fef5ffbf3102119/lib/api/experiences.ts
import { apiDelete, apiFetch, apiPatch, apiPost, apiPut } from './client';

export type ExperienceKind =
  'work' | 'internship' | 'project' | 'research' | 'campus' | 'volunteer' | 'other';
export type ExperienceStatus = 'draft' | 'ready' | 'archived';
export type ExperienceListStatus = 'active' | ExperienceStatus;
export type ExperienceSort = 'updated_at_desc' | 'created_at_desc' | 'created_at_asc';

export interface EvidenceItem {
  id: number;
  action: string;
  result: string | null;
  metrics: string | null;
  created_at: string;
  updated_at: string;
}

export interface ExperienceRead {
  experience_id: number;
  kind: ExperienceKind;
  title: string;
  organization: string | null;
  role: string | null;
  location: string | null;
  start_date: string | null;
  end_date: string | null;
  is_current: boolean;
  background: string | null;
  evidence_ids: number[];
  technologies: string[];
  tags: string[];
  notes: string | null;
  status: ExperienceStatus;
  completeness: number;
  archived_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface ExperienceDetail extends ExperienceRead {
  evidence_items: EvidenceItem[];
  missing_dimensions: string[];
  suggested_questions: string[];
  field_states: ExperienceFieldState[];
}

export interface ExperienceFieldState {
  key: string;
  ref_id: number | null;
  status: 'complete' | 'incomplete';
  revision: number;
}

export interface ExperienceCreate {
  kind?: ExperienceKind;
  title?: string;
  organization?: string | null;
  role?: string | null;
  location?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  is_current?: boolean;
  background?: string | null;
  technologies?: string[];
  tags?: string[];
  notes?: string | null;
}

export interface ExperienceUpdate extends ExperienceCreate {
  expected_field_revisions?: Record<string, number>;
}

export interface EvidenceCreate {
  action: string;
  result?: string | null;
  metrics?: string | null;
}

export interface EvidenceUpdate {
  action?: string | null;
  result?: string | null;
  metrics?: string | null;
  expected_revision?: number;
}

export interface ExperienceEvidenceSave {
  evidence_id: number;
  action: string;
  result: string | null;
  metrics: string | null;
  expected_revision: number;
}

export interface ExperienceGlobalSave {
  experience: ExperienceUpdate;
  evidence_items: ExperienceEvidenceSave[];
  new_evidence: EvidenceCreate | null;
  expected_collection_revision: number;
}

export interface ExperienceEvidenceSave {
  evidence_id: number;
  action: string;
  result: string | null;
  metrics: string | null;
  expected_revision: number;
}

export interface ExperienceGlobalSave {
  experience: ExperienceUpdate;
  evidence_items: ExperienceEvidenceSave[];
  new_evidence: EvidenceCreate | null;
  expected_collection_revision: number;
}

export interface ExperienceListQuery {
  q?: string;
  kind?: ExperienceKind;
  status?: ExperienceListStatus;
  sort?: ExperienceSort;
}

export interface ExperienceListResponse {
  items: ExperienceRead[];
  total: number;
}

export interface DeletionImpactResponse {
  affected_matches: Array<{ match_id: number; job_title: string }>;
  affected_resumes: string[];
}

export interface ReadyConflictResponse {
  completeness: number;
  missing_dimensions: string[];
}

export class ExperienceReadyConflictError extends Error {
  constructor(public readonly conflict: ReadyConflictResponse) {
    super('Experience is not complete enough to mark ready');
    this.name = 'ExperienceReadyConflictError';
  }
}

function responseErrorMessage(payload: unknown): string | null {
  if (!payload || typeof payload !== 'object') return null;
  const detail = (payload as { detail?: unknown }).detail;
  if (typeof detail === 'string') return detail;
  if (Array.isArray(detail)) {
    const messages = detail
      .map((item) =>
        item && typeof item === 'object' && 'msg' in item
          ? String((item as { msg: unknown }).msg)
          : null
      )
      .filter((message): message is string => Boolean(message));
    return messages.length ? messages.join('; ') : null;
  }
  if (detail && typeof detail === 'object') {
    try {
      return JSON.stringify(detail);
    } catch {
      return null;
    }
  }
  return null;
}

async function parseResponse<T>(response: Response, fallback: string): Promise<T> {
  const text = await response.text();
  let payload: unknown = null;
  if (text) {
    try {
      payload = JSON.parse(text);
    } catch {
      payload = null;
    }
  }
  if (!response.ok) {
    throw new Error(
      responseErrorMessage(payload) || text || `${fallback} (status ${response.status}).`
    );
  }
  return payload as T;
}

function experiencePath(experienceId: number): string {
  return `/experiences/${encodeURIComponent(String(experienceId))}`;
}

export async function listExperiences(
  query: ExperienceListQuery = {},
  signal?: AbortSignal
): Promise<ExperienceListResponse> {
  const params = new URLSearchParams();
  if (query.q) params.set('q', query.q);
  if (query.kind) params.set('kind', query.kind);
  if (query.status) params.set('status', query.status);
  if (query.sort) params.set('sort', query.sort);
  const suffix = params.size ? `?${params.toString()}` : '';
  return parseResponse<ExperienceListResponse>(
    await apiFetch(`/experiences${suffix}`, signal ? { signal } : undefined),
    'Failed to load experiences'
  );
}

export async function importExperienceText(text: string): Promise<ExperienceDetail> {
  return parseResponse<ExperienceDetail>(
    await apiPost('/experiences/import-text', { text }),
    'Failed to import experience text'
  );
}

export async function createExperience(payload: ExperienceCreate): Promise<ExperienceDetail> {
  return parseResponse<ExperienceDetail>(
    await apiPost('/experiences', payload),
    'Failed to create experience'
  );
}

export async function fetchExperience(
  experienceId: number,
  signal?: AbortSignal
): Promise<ExperienceDetail> {
  return parseResponse<ExperienceDetail>(
    await apiFetch(experiencePath(experienceId), signal ? { signal } : undefined),
    'Failed to load experience'
  );
}

export async function patchExperience(
  experienceId: number,
  payload: ExperienceUpdate
): Promise<ExperienceDetail> {
  return parseResponse<ExperienceDetail>(
    await apiPatch(experiencePath(experienceId), payload),
    'Failed to update experience'
  );
}

export async function saveExperience(
  experienceId: number,
  payload: ExperienceGlobalSave
): Promise<ExperienceDetail> {
  return parseResponse<ExperienceDetail>(
    await apiPut(`${experiencePath(experienceId)}/save`, payload),
    'Failed to save experience'
  );
}

export async function markExperienceReady(experienceId: number): Promise<ExperienceDetail> {
  const response = await apiPost(`${experiencePath(experienceId)}/mark-ready`, {});
  if (response.status === 409) {
    const payload = (await response.json()) as ReadyConflictResponse;
    throw new ExperienceReadyConflictError(payload);
  }
  return parseResponse<ExperienceDetail>(response, 'Failed to mark experience ready');
}

export async function archiveExperience(experienceId: number): Promise<ExperienceDetail> {
  return parseResponse<ExperienceDetail>(
    await apiPost(`${experiencePath(experienceId)}/archive`, {}),
    'Failed to archive experience'
  );
}

export async function restoreExperience(experienceId: number): Promise<ExperienceDetail> {
  return parseResponse<ExperienceDetail>(
    await apiPost(`${experiencePath(experienceId)}/restore`, {}),
    'Failed to restore experience'
  );
}

export async function getDeletionImpact(
  experienceId: number,
  signal?: AbortSignal
): Promise<DeletionImpactResponse> {
  return parseResponse<DeletionImpactResponse>(
    await apiFetch(
      `${experiencePath(experienceId)}/deletion-impact`,
      signal ? { signal } : undefined
    ),
    'Failed to load deletion impact'
  );
}

export async function deleteExperiencePermanently(experienceId: number): Promise<void> {
  await parseResponse<unknown>(
    await apiDelete(`${experiencePath(experienceId)}/permanent`),
    'Failed to permanently delete experience'
  );
}

export async function createEvidence(
  experienceId: number,
  payload: EvidenceCreate
): Promise<ExperienceDetail> {
  return parseResponse<ExperienceDetail>(
    await apiPost(`${experiencePath(experienceId)}/evidence`, payload),
    'Failed to create evidence'
  );
}

export async function patchEvidence(
  experienceId: number,
  evidenceId: number,
  payload: EvidenceUpdate
): Promise<ExperienceDetail> {
  return parseResponse<ExperienceDetail>(
    await apiPatch(
      `${experiencePath(experienceId)}/evidence/${encodeURIComponent(String(evidenceId))}`,
      payload
    ),
    'Failed to update evidence'
  );
}

export async function deleteEvidence(
  experienceId: number,
  evidenceId: number
): Promise<ExperienceDetail> {
  return parseResponse<ExperienceDetail>(
    await apiDelete(
      `${experiencePath(experienceId)}/evidence/${encodeURIComponent(String(evidenceId))}`
    ),
    'Failed to delete evidence'
  );
}

export async function reorderEvidence(
  experienceId: number,
  evidenceIds: number[]
): Promise<ExperienceDetail> {
  return parseResponse<ExperienceDetail>(
    await apiPut(`${experiencePath(experienceId)}/evidence-order`, { evidence_ids: evidenceIds }),
    'Failed to reorder evidence'
  );
}

File: apps/backend/app/experience/schemas/experiences.py
"""个人经历库的 Pydantic 契约。"""

import re
from enum import Enum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.experience.schemas.evidence_items import EvidenceCreate, EvidenceRead


class ExperienceKind(str, Enum):
    work = "work"
    internship = "internship"
    project = "project"
    research = "research"
    campus = "campus"
    volunteer = "volunteer"
    other = "other"


class ExperienceStatus(str, Enum):
    draft = "draft"
    ready = "ready"
    archived = "archived"


_YEAR_MONTH_RE = re.compile(r"^\d{4}-(0[1-9]|1[0-2])$")
MAX_IMPORT_TEXT_LENGTH = 20_000


def _normalize_labels(value: object) -> list[str]:
    if value is None:
        return []
    if not isinstance(value, list):
        return value  # type: ignore[return-value]

    normalized: list[str] = []
    seen: set[str] = set()
    for item in value:
        if not isinstance(item, str):
            normalized.append(item)
            continue
        label = item.strip()
        key = label.casefold()
        if label and key not in seen:
            normalized.append(label)
            seen.add(key)
    return normalized


# BaseModel：Pydantic 提供的模型基类，用来定义带有类型检查、数据校验和转换能力的数据结构。
class _ExperienceWritable(BaseModel):
    """创建与局部更新共用的可编辑字段。"""

    kind: ExperienceKind | None = None
    title: str | None = None
    organization: str | None = None
    role: str | None = None
    location: str | None = None
    start_date: str | None = None
    end_date: str | None = None
    is_current: bool | None = None
    background: str | None = None
    technologies: list[str] | None = None
    tags: list[str] | None = None
    notes: str | None = None

    @field_validator("start_date", "end_date")
    @classmethod
    def _validate_year_month(cls, value: str | None) -> str | None:
        if value is not None and not _YEAR_MONTH_RE.fullmatch(value):
            raise ValueError("date must use YYYY-MM format")
        return value

    @field_validator("technologies", "tags", mode="before")
    @classmethod
    def _normalize_label_lists(cls, value: object) -> list[str] | None:
        if value is None:
            return None
        return _normalize_labels(value)

    @model_validator(mode="after")
    def _current_experience_cannot_end(self) -> "_ExperienceWritable":
        if self.is_current is True and self.end_date is not None:
            raise ValueError("current experiences cannot have an end_date")
        return self


class ExperienceCreate(_ExperienceWritable):
    """客户端手动创建经历时提供的字段。"""

    model_config = ConfigDict(
        extra="forbid"
    )  # model_config：Pydantic配置的固定名称，extra="forbid"：如果输入数据包含模型中没有定义的额外字段，就直接校验失败。

    kind: ExperienceKind = ExperienceKind.other
    title: str = ""
    is_current: bool = False
    technologies: list[str] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)


class ExperienceUpdate(_ExperienceWritable):
    """客户端提供的局部更新；服务端字段保持不可写。"""

    model_config = ConfigDict(extra="forbid")

    expected_field_revisions: dict[str, int] = Field(default_factory=dict)


class ExperienceEvidenceSave(EvidenceCreate):
    """全局保存中的 Evidence；有 ID 更新，无 ID 创建。"""

    model_config = ConfigDict(extra="forbid")

    evidence_id: int | None = Field(default=None, gt=0)
    expected_revision: int | None = Field(default=None, ge=0)

    @model_validator(mode="after")
    def _existing_evidence_requires_revision(self) -> "ExperienceEvidenceSave":
        if (self.evidence_id is None) != (self.expected_revision is None):
            raise ValueError(
                "existing evidence requires both evidence_id and expected_revision"
            )
        return self


class ExperienceGlobalSave(BaseModel):
    """聚合保存：有 experience_id 覆盖，无 experience_id 创建。"""

    model_config = ConfigDict(extra="forbid")

    experience_id: int | None = Field(default=None, gt=0)
    experience: ExperienceUpdate
    evidence_items: list[ExperienceEvidenceSave] = Field(default_factory=list)
    expected_collection_revision: int | None = Field(default=None, ge=0)

    @model_validator(mode="after")
    def _validate_create_or_update_contract(self) -> "ExperienceGlobalSave":
        if self.experience_id is None:
            if self.expected_collection_revision is not None:
                raise ValueError(
                    "new experience must not provide a collection revision"
                )
            if self.experience.expected_field_revisions:
                raise ValueError("new experience must not provide field revisions")
            if any(item.evidence_id is not None for item in self.evidence_items):
                raise ValueError("new experience cannot reference existing evidence")
        elif self.expected_collection_revision is None:
            raise ValueError("existing experience requires a collection revision")
        return self


class ExperienceRead(BaseModel):
    """返回给客户端的已存储经历字段。"""

    experience_id: int
    kind: ExperienceKind
    title: str
    organization: str | None = None
    role: str | None = None
    location: str | None = None
    start_date: str | None = None
    end_date: str | None = None
    is_current: bool = False
    background: str | None = None
    evidence_ids: list[int] = Field(default_factory=list)
    technologies: list[str] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
    notes: str | None = None
    status: ExperienceStatus
    completeness: int = Field(ge=0, le=100)
    archived_at: str | None = None
    created_at: str
    updated_at: str


class ExperienceDetail(ExperienceRead):
    """包含展开证据和重算引导信息的经历。"""

    evidence_items: list[EvidenceRead] = Field(default_factory=list)
    missing_dimensions: list[str] = Field(default_factory=list)
    suggested_questions: list[str] = Field(default_factory=list)
    field_states: list["ExperienceFieldStateRead"] = Field(default_factory=list)


class ExperienceFieldStateRead(BaseModel):
    """返回给经历编辑器的一项字段状态和修订号。"""

    key: str
    ref_id: int | None = None
    status: Literal["complete", "incomplete"]
    revision: int = Field(ge=0)


class ExperienceListQuery(BaseModel):
    """本地经历库支持的筛选和排序方式。"""

    model_config = ConfigDict(extra="forbid")

    q: str | None = None
    kind: ExperienceKind | None = None
    status: Literal["active", "draft", "ready", "archived"] = "active"
    sort: Literal["updated_at_desc", "created_at_desc", "created_at_asc"] = (
        "updated_at_desc"
    )


class ExperienceListResponse(BaseModel):
    """为后续分页预留结构且不改变项目契约的列表响应。"""

    items: list[ExperienceRead] = Field(default_factory=list)
    total: int = Field(ge=0, default=0)


class ExperienceCompleteness(BaseModel):
    """派生且不持久化的完整度引导信息。"""

    completeness: int = Field(ge=0, le=100)
    missing_dimensions: list[str] = Field(default_factory=list)
    suggested_questions: list[str] = Field(default_factory=list)


class ExperienceImportTextRequest(BaseModel):
    """仅用于一次结构化导入请求的临时源文本。"""

    model_config = ConfigDict(extra="forbid")

    text: str = Field(min_length=1, max_length=MAX_IMPORT_TEXT_LENGTH)

    @field_validator("text")
    @classmethod
    def _reject_blank_text(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("text must not be blank")
        return value


class ReadyConflictResponse(BaseModel):
    """标记就绪被拒绝时返回的当前分数和缺失事实。"""

    completeness: int = Field(ge=0, le=100)
    missing_dimensions: list[str] = Field(default_factory=list)


class DeletionImpactMatch(BaseModel):
    """受永久删除影响的一条未来匹配记录。"""

    match_id: int
    job_title: str


class DeletionImpactResponse(BaseModel):
    """用于永久删除审查的稳定、向前兼容影响契约。"""

    affected_matches: list[DeletionImpactMatch] = Field(default_factory=list)
    affected_resumes: list[str] = Field(default_factory=list)
